wire-compiler/src/main/java/com/squareup/wire/WireCompiler.java

```diff
@@ -1254,4 +1269,59 @@ private boolean javaTypeIsComplete(String type) {
   private String fullyQualifiedJavaName(MessageElement messageType, String type) {
     return TypeInfo.isScalar(type) ? null : javaName(fullyQualifiedName(messageType, type));
   }
+
+  private void compileWithJavaPoet() throws WireException {
+    Set<String> parsedFiles = new LinkedHashSet<String>();
+    Loader loader = new Loader(repoPath, io);
+    for (String sourceFilename : options.sourceFileNames) {
+      String sourcePath = repoPath + File.separator + sourceFilename;
+      parsedFiles.add(sourcePath);
+      try {
+        loader.add(sourceFilename);
+      } catch (IOException e) {
+        throw new WireException("Error loading symbols for " + sourcePath, e);
+      }
+    }
+
+    List<WireProtoFile> wireProtoFiles = loader.loaded();
+    Linker linker = new Linker();
+    linker.link(wireProtoFiles);
+
+    if (!typesToEmit.isEmpty()) {
+      log.info("Analyzing dependencies of root types.");
+      wireProtoFiles = new Pruner().retainRoots(wireProtoFiles, typesToEmit);
+    }
+
+    JavaCompiler javaCompiler = JavaCompiler.get(wireProtoFiles);
+    for (WireProtoFile wireProtoFile : wireProtoFiles) {
+      if (!parsedFiles.contains(wireProtoFile.sourcePath())) {
+        continue; // Don't emit anything for files not explicitly compiled.
+      }
+
+      for (com.squareup.wire.model.WireType type : wireProtoFile.types()) {
```

Comment:
I currently have a collision with the other `WireType` class. I'm going to probably need a different suffix for these model classes; that `WireType` is public API in the runtime.